google/cloud/spanner/timestamp.cc

```diff
@@ -52,347 +40,64 @@ Status NegativeOverflow(std::string const& type) {
-// subsecond ratios are powers of 10.
-StatusOr<Timestamp> Timestamp::FromRatio(std::intmax_t const count,
-                                         std::intmax_t const numerator,
-                                         std::intmax_t const denominator) {
-  constexpr auto kDestType = "google::cloud::spanner::Timestamp";
-
-  auto sec = count / denominator;
-  if (sec >= 0) {
-    if (sec > std::numeric_limits<std::intmax_t>::max() / numerator) {
-      return PositiveOverflow(kDestType);
-    }
-  } else {
-    if (sec < std::numeric_limits<std::intmax_t>::min() / numerator) {
-      return NegativeOverflow(kDestType);
-    }
-  }
-  sec *= numerator;
-
-  auto sscount = count - sec * denominator / numerator;
-  if (denominator > kNanosPerSecond) {
-    auto const divider = denominator / kNanosPerSecond;
-    if (sscount < 0) sscount -= divider - 1;  // floor
-    return FromCounts(sec, sscount / divider * numerator);
-  }
-  auto const multiplier = kNanosPerSecond / denominator;
-  return FromCounts(sec, sscount * multiplier * numerator);
-}
-
-// [sec, nsec] => bounded (count * numerator/denominator) seconds.
-//
-// Only designed to handle the ratios of the std::chrono::duration helper
-// aliases, where either the numerator or the denominator is 1, and where
-// subsecond ratios are powers of 10.
-StatusOr<std::intmax_t> Timestamp::ToRatio(
-    std::intmax_t const min, std::intmax_t const max,
-    std::intmax_t const numerator, std::intmax_t const denominator) const {
-  constexpr auto kDestType = "std::chrono::time_point";
-
-  auto count = sec_ / numerator;
-  if (count >= 0) {
-    if (count > std::numeric_limits<std::intmax_t>::max() / denominator) {
-      return PositiveOverflow(kDestType);
-    }
-  } else {
-    if (count < std::numeric_limits<std::intmax_t>::min() / denominator) {
-      // Might be premature to declare overflow on an intermediate value.
-      return NegativeOverflow(kDestType);
-    }
-  }
-  count *= denominator;
-
-  auto ncount = nsec_ / numerator;
-  if (denominator < kNanosPerSecond) {
-    ncount /= kNanosPerSecond / denominator;
-  } else {
-    ncount *= denominator / kNanosPerSecond;
-  }
-
-  if (count > std::numeric_limits<std::intmax_t>::max() - ncount) {
-    return PositiveOverflow(kDestType);
-  }
-  count += ncount;
-
-  if (count > max) return PositiveOverflow(kDestType);
-  if (count < min) return NegativeOverflow(kDestType);
-  return count;
+std::ostream& operator<<(std::ostream& os, Timestamp ts) {
+  return os << internal::TimestampToRFC3339(ts);
 }
 
 namespace internal {
 
+// Timestamp objects are always formatted in UTC, and we always format them
+// with a trailing 'Z'. However, we're a bit more liberal in the UTC offsets we
+// accept, thus the use of '%Ez' in kParseSpec.
+auto constexpr kFormatSpec = "%E4Y-%m-%dT%H:%M:%E*SZ";
+auto constexpr kParseSpec = "%Y-%m-%dT%H:%M:%E*S%Ez";
+
 StatusOr<Timestamp> TimestampFromRFC3339(std::string const& s) {
```

Comment:
It is a shame that we lost micro-benchmark coverage for the `string` <=> `Timestamp` conversions, but I guess the performance is now out of our hands.